### PR TITLE
StorageWriter retries

### LINKED ISSUE #119: IPFS Upload Retries
needs #112

### PR DESCRIPTION
**Pull request by kennylavender**
[PR Process](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-process) - [PR Review Checklist](https://github.com/poetapp/documentation/blob/master/process/pr-review.md#pr-review-checklist)

## Description of Changes

closes #119 

Converts StorageWriter to upload claims to IPFS asynchronously with retries.

* Add Service for upload intervals
* Add `uploadClaimIntervalInSeconds` and `uploadClaimMaxAttempts` configuration settings
* Adjust Router.ts and Messaging to support the new interval
* Removed brittle new/instance of unit tests ( I might undo this, and maybe redo in another PR )
* Split out DB functionality into a DOA to simplify the Claim Controller. Do a better job of `program to an interface, not an implementation` by defining the Database interface separately from the implementation and by removing details of mongo from the interface definition.
* Adjusted the ClaimController to handle the new logic. I tried to really split the functions into 'do one thing' functions including error handling and make use of pipeP. I feel like there is still room for improvement. It feels difficult to come up with a good format around classes, arrow functions, and pipeP. 
* Put logging in the ClaimControler into pipeP, I think this gets rid of lots of the noise of having the repeated logging setup in the main functions of the ClaimController. Let me know if you don't like this @lautarodragan and I will revert it, or ideas to improve it welcome as well.
* Remove nondeterministic test `gettingThePostedWorkShouldRetrieveIPFSHash`. The test was posting a work, then fetching the work and checking for an ipfsFileHash in a timestamp property. This worked while storage was synchronous, but now with storage being asynchronous, it will not work unless we add a huge delay. The reason it won't work is that our tests all use the same database instance and we are posting many works to the node, so the works will stack up in the storage claims collection and all be awaiting being uploaded to ipfs which is why this test will not work without a huge delay, it would require waiting for all those posted works before it to be uploaded to ipfs first.
* moved interfaces to the same file as their classes, it helped solve the conflicts with master easier.
* use custom exception and instanceOf

### TSLint
- TSLint complains about public arrow functions being after private arrow functions in Classes. This is a requirement though of arrow function use in Classes.
- TSLint complains about public functions being created with pipeP being placed after private properties and attempts to move them above the constructor and private properties.

~I will fix these TSLint issues in another PR either before or after merging this PR, for now just using comments to disable its complaints.~

PR to fix here: https://github.com/poetapp/tslint-rules/pull/129

### CODE REVIEW

**Comment by kennylavender** [comment]
Getting errors in Travis about not being able to connect to bitcoind service. Nothing has changed in this PR around the package-lock.json, package.json, travis.yml or docker configuration that should be causing this.

@lautarodragan any ideas?

```
[2018-09-05T19:18:16.667Z] ERROR (5364 on travis-job-7b02583e-a840-4bc9-9ef4-0749bf70eafb): Uncaught Error
    module: "BlockchainReader"
    file: "BlockchainReaderService"
    method: "scanNewBlock"
    error: {
      "cause": {
        "errno": "ECONNREFUSED",
        "code": "ECONNREFUSED",
        "syscall": "connect",
        "address": "127.0.0.1",
        "port": 18332,
        "stack": [
          "Error: connect ECONNREFUSED 127.0.0.1:18332",
          "    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1161:14)"
        ],
        "message": "connect ECONNREFUSED 127.0.0.1:18332",
        "type": "Error"
      },
      "isOperational": true,
      "errno": "ECONNREFUSED",
      "code": "ECONNREFUSED",
      "syscall": "connect",
      "address": "127.0.0.1",
      "port": 18332,
      "stack": [
        "Error: connect ECONNREFUSED 127.0.0.1:18332",
        "    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1161:14)"
      ],
      "message": "connect ECONNREFUSED 127.0.0.1:18332",
      "type": "OperationalError"
    }
    blockHeight: 1279551
    lastBlockHeight: 1279550
no-unused-variable is deprecated. Since TypeScript 2.9. Please use the built-in compiler checks instead.
[2018-09-05T19:18:21.636Z] ERROR (5364 on travis-job-7b02583e-a840-4bc9-9ef4-0749bf70eafb): Uncaught Error
    module: "BlockchainReader"
    file: "BlockchainReaderService"
    method: "scanNewBlock"
    error: {
      "cause": {
        "errno": "ECONNREFUSED",
        "code": "ECONNREFUSED",
        "syscall": "connect",
        "address": "127.0.0.1",
        "port": 18332,
        "stack": [
          "Error: connect ECONNREFUSED 127.0.0.1:18332",
          "    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1161:14)"
        ],
        "message": "connect ECONNREFUSED 127.0.0.1:18332",
        "type": "Error"
      },
      "isOperational": true,
      "errno": "ECONNREFUSED",
      "code": "ECONNREFUSED",
      "syscall": "connect",
      "address": "127.0.0.1",
      "port": 18332,
      "stack": [
        "Error: connect ECONNREFUSED 127.0.0.1:18332",
        "    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1161:14)"
      ],
      "message": "connect ECONNREFUSED 127.0.0.1:18332",
      "type": "OperationalError"
    }
    blockHeight: 1279551
    lastBlockHeight: 1279550
```

**Comment by lautarodragan** [comment]
Ugh, sorry about that. Don't know why tests were passing in my PR, but I never updated .travis.yml to work with the new config. I'll work on a fix

**Comment by kennylavender** [comment]
@lautarodragan and @poetapp/quality-engineering ready for review.

File: src/StorageWriter/DatabaseMongo.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  public readonly addError = async (error: object) => {
+    await this.errors.insertOne(error)
+  }
+
+  public readonly addClaim = async (claim: Claim) => {
+    await this.claims.insertOne({ claimId: view(L.id, claim), claim, storageAttempts: 0, ipfsFileHash: null })
+  }
+
+  public readonly addClaimHash = async (claimId: string, ipfsFileHash: string) => {
```
Comment (wzalazar) [inline src/StorageWriter/DatabaseMongo.ts]: do we need `try/catch`? what's happen if there are any error

File: src/StorageWriter/Errors.test.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+      given: 'all arguments',
+      should: 'return the correct object',
+      expected: {
+        type: UPLOAD_CLAIM_ERROR,
+        details: {
+          message,
+          claim,
+        }
+      },
+      actual: uploadClaimError(message, claim)
```
[the remark this answers is not in the document]
Comment (kennylavender) [reply src/StorageWriter/Errors.test.ts]: No reason, will switch

File: src/Configuration.ts [hunk 3]
```diff
@@ -81,6 +83,8 @@ const defaultConfiguration: Configuration = {
 
   readDirectoryIntervalInSeconds: 30,
 
+  uploadClaimIntervalInSeconds: 30,
```
[the remark this answers is not in the document]
Comment (kennylavender) [reply src/Configuration.ts]: I think your thinking this is an interface, but it's not an interface, it's a JavaScript object if you expand the code :)

**Comment by lautarodragan** [comment]
@kennylavender actually, even though the lack of bitcoind instance in Travis is causing errors, the only failing test doesn't depend on it: gettingThePostedWorkShouldRetrieveIPFSHash.

File: src/Configuration.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (wzalazar) [reply src/Configuration.ts]: oh, yeah you're right, my bad

File: src/StorageWriter/ClaimController.ts [hunk 4]
```diff
@@ -1,51 +1,111 @@
...
+    try {
+      const ipfsFileHash = await this.uploadClaim(claim)
+      return set(L.ipfsFileHash, ipfsFileHash, data)
+    } catch (error) {
+      await this.handleStoreClaimError(error, claim)
+    }
+  }
+
+  private readonly handleAddIPFSHashToClaimError = async (error: Error) => {
+    throw new Error('Failed to update claim hash')
```
[the remark this answers is not in the document]
Comment (kennylavender) [reply src/StorageWriter/ClaimController.ts]: Good point, this would be a good error to also store in the mongo error collection, I will add it.

File: src/StorageWriter/DatabaseMongo.ts (hunk 1 above)
Comment (kennylavender) [reply src/StorageWriter/DatabaseMongo.ts]: This is being caught from the Claim Controller now

**Comment by kennylavender** [comment]
@lautarodragan thanks, I didn't notice the tests even successfully ran, I will take a look.

File: src/StorageWriter/DatabaseMongo.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import { lensProp, lensPath, view, isNil, pipeP } from 'ramda'
+
+import { Database } from './Database'
+
+const L = {
+  id: lensProp('id'),
+  valueClaim: lensPath(['value', 'claim']),
+}
+
+const MAX_STORAGE_ATTEMPTS = 10
```
Comment (lautarodragan) [inline src/StorageWriter/DatabaseMongo.ts]: configurable?

File: src/StorageWriter/DatabaseMongo.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      },
+      {
+        $inc: { storageAttempts: 1 },
+        $set: { lastStorageAttemptTime: new Date().getTime() },
+      }
+    )
+    return view(L.valueClaim, response)
+  }
+
+  private readonly handleNoClaimsFound = async (claim: Claim) => {
```
Comment (lautarodragan) [inline src/StorageWriter/DatabaseMongo.ts]: Is this function DB related?

File: src/StorageWriter/Errors.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import { Claim } from '@po.et/poet-js'
+
+export const UPLOAD_CLAIM_ERROR = 'UPLOAD_CLAIM_ERROR'
```
Comment (lautarodragan) [inline src/StorageWriter/Errors.ts]: Can you turn these into a string enum?

File: src/StorageWriter/ClaimController.ts [hunk 8]
```diff
@@ -1,51 +1,112 @@
...
+    try {
+      await this.db.addClaimHash(view(L.claimId, data), view(L.ipfsFileHash, data))
+      return data
+    } catch (error) {
+      await this.handleAddIPFSHashToClaimError(error, view(L.claim, data), view(L.ipfsFileHash, data))
+    }
+  }
+
+  // tslint:disable-next-line
+  public storeNextClaim = pipeP(
```
Comment (lautarodragan) [inline src/StorageWriter/ClaimController.ts]: Love this

File: src/StorageWriter/ClaimController.ts [hunk 9]
```diff
@@ -1,51 +1,112 @@
...
     this.logger = childWithFileName(logger, __filename)
     this.db = db
-    this.collection = this.db.collection('storageWriter')
-    this.messaging = messaging
     this.ipfs = ipfs
   }
 
-  async create(claim: Claim): Promise<void> {
-    const logger = this.logger.child({ method: 'create' })
+  private readonly log = (level: LogTypes) => (message: string) => async (value: any) => {
```
Comment (lautarodragan) [inline src/StorageWriter/ClaimController.ts]: Move to LogHelper? Might want to reuse this in other places

**Comment by kennylavender** [comment]
@lautarodragan @poetapp/quality-engineering I had to remove the failing test because it was nondeterministic without a huge delay. Updated description above ( last item on the change list ).

File: src/StorageWriter/ClaimController.ts (hunk 9 above)
Comment (kennylavender) [reply src/StorageWriter/ClaimController.ts]: Sounds good to me. You ok with me doing this in a separate PR?

Comment (lautarodragan) [reply src/StorageWriter/ClaimController.ts]: Yep

**Comment by lautarodragan** [comment]
That's OK for this PR. We'll improve our functional tests very soon in other PRs.

File: src/StorageWriter/DatabaseMongo.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+import { DatabaseMongoConfiguration } from './DatabaseMongoConfiguration'
+
+const L = {
+  valueClaim: lensPath(['value', 'claim']),
+}
+
+export const getClaimFromFindAndUpdateResponse = (response: FindAndModifyWriteOpResultObject): Claim | undefined =>
+  view(L.valueClaim, response)
+
+export const throwIfClaimNotFound = (claim: Claim): Claim => {
```
Comment (kennylavender) [inline src/StorageWriter/DatabaseMongo.ts]: @lautarodragan are you good with how I've removed this function from the class so it could be unit tested?

**Comment by kennylavender** [comment]
@lautarodragan @poetapp/quality-engineering 

Ready for another round of reviews. I don't think this will require full manual testing to be merged. It will only require testing up to the point of publishing the ClaimIPFSHash rabbitmq message.

File: src/StorageWriter/Database.ts [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import { Claim } from '@po.et/poet-js'
+
+export interface Database {
+  addClaim: (claim: Claim) => Promise<void>
```
Comment (lautarodragan) [inline src/StorageWriter/Database.ts]: readonly?

File: src/StorageWriter/DatabaseMongo.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+      {
+        $and: [{ ipfsFileHash: null }, { storageAttempts: { $lt: this.maxStorageAttempts } }],
+      },
+      {
+        $inc: { storageAttempts: 1 },
+        $set: { lastStorageAttemptTime: new Date().getTime() },
+      }
+    )
+
+  // tslint:disable-next-line
```
[the remark this answers is not in the document]
Comment (kennylavender) [reply src/StorageWriter/DatabaseMongo.ts]: I haven't found the rule that does this, and haven't had the time to keep looking yet.

File: src/StorageWriter/DatabaseMongo.ts [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+  public readonly start = async () => {
+    await this.claims.createIndex({ claimId: 1 }, { unique: true })
+  }
+
+  public readonly addError = async (error: object) => {
+    await this.errors.insertOne(error)
+  }
+
+  public readonly addClaim = async (claim: Claim) => {
+    await this.claims.insertOne({ claimId: claim.id, claim, storageAttempts: 0, ipfsFileHash: null })
```
[the remark this answers is not in the document]
Comment (kennylavender) [reply src/StorageWriter/DatabaseMongo.ts]: Don't think so, I tried it but it wouldn't allow me, I could have done something wrong though.

File: src/StorageWriter/DatabaseMongo.ts [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+import { isNil, pipeP, lensPath, view } from 'ramda'
+
+import { Database } from './Database'
+import { DatabaseMongoConfiguration } from './DatabaseMongoConfiguration'
+
+const L = {
+  valueClaim: lensPath(['value', 'claim']),
+}
+
+export const getClaimFromFindAndUpdateResponse = (response: FindAndModifyWriteOpResultObject): Claim | undefined =>
```
[the remark this answers is not in the document]
Comment (kennylavender) [reply src/StorageWriter/DatabaseMongo.ts]: The benefit here is that lenses will return undefined without needing ternaries for checking nested properties.

I don't believe manually getting the value from the response will provide anymore type safety as `response.value` is already an 'any' type.

File: src/StorageWriter/DatabaseMongo.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (lautarodragan) [reply src/StorageWriter/DatabaseMongo.ts]: I have the vague idea that I have used them in the past, but not sure.

Should be possible according to [the docs](https://docs.mongodb.com/manual/core/index-single/#index-embedded-documents).

I'll give it a try and see what happens

File: src/StorageWriter/DatabaseMongo.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (lautarodragan) [reply src/StorageWriter/DatabaseMongo.ts]: It's [member-ordering](https://palantir.github.io/tslint/rules/member-ordering/) according to my IDE. 

Setting `tslint.json` like this seems to do the trick:

```json
{
  "extends": [
    "@po.et/tslint-rules"
  ],
  "rules": {
    "member-ordering": false
  }
}
```

Although I'd rather tweak it a bit if possible, setting a more specific value that doesn't conflict with our current use case but not disabling completely.

File: src/StorageWriter/ClaimController.ts [hunk 15]
```diff
@@ -1,51 +1,102 @@
...
 
 import { childWithFileName } from 'Helpers/Logging'
-import { Exchange } from 'Messaging/Messages'
-import { Messaging } from 'Messaging/Messaging'
 
+import { Database } from './Database'
+import { uploadClaimError, updateClaimHashError } from './Errors'
 import { IPFS } from './IPFS'
 
+enum LogTypes {
```
[the remark this answers is not in the document]
Comment (kennylavender) [reply src/StorageWriter/ClaimController.ts]: Fixed!

File: src/StorageWriter/DatabaseMongo.ts [hunk 16]
```diff
@@ -0,0 +1,66 @@
+import { Claim } from '@po.et/poet-js'
```
Comment (lautarodragan) [inline src/StorageWriter/DatabaseMongo.ts]: Weren't we calling these WhateverDAO to avoid conflicting with Mongo's Database? (Just noticed BatchWriter's FileDAO)

Comment (kennylavender) [reply src/StorageWriter/DatabaseMongo.ts]: We were calling them by that format, but not because of conflicts with Mongo's naming. The reason was that each DAO was dealing with a single collection. But that means that the claim controller needs to know about both collections independently. I thought it might be better for it to just deal with a single database interface, but I could go either way on this 🤷‍♂️

**Comment by kennylavender** [comment]
Todo: undo Database interface extraction

File: src/StorageWriter/ClaimController.ts [hunk 17]
```diff
@@ -1,51 +1,102 @@
...
+  private readonly findNextClaim = async (): Promise<StoreNextClaimData> => {
+    const claim = await this.db.findNextClaim()
+    return { claim }
+  }
+
+  private readonly uploadClaim = (claim: Claim) => this.ipfs.addText(JSON.stringify(claim))
+
+  private readonly handleStoreClaimError = async (error: Error, claim: Claim) => {
+    await this.db.addError(uploadClaimError(error.message, claim))
+    throw new Error('Failed to store claim')
```
[the remark this answers is not in the document]
Comment (kennylavender) [reply src/StorageWriter/ClaimController.ts]: I was thinking it would be better to track each attempt error because it might fail for different reasons.

File: src/StorageWriter/ClaimController.ts [hunk 18]
```diff
@@ -1,51 +1,102 @@
...
+    await this.db.addError(updateClaimHashError(error.message, claim, hash))
+    throw new Error('Failed to update claim hash')
+  }
+
+  private readonly addIPFSHashToClaim = async (data: StoreNextClaimData): Promise<StoreNextClaimData> => {
+    const { claim, ipfsFileHash } = data
+    try {
+      await this.db.addClaimHash(claim.id, ipfsFileHash)
+      return data
+    } catch (error) {
```
[the remark this answers is not in the document]
Comment (kennylavender) [reply src/StorageWriter/ClaimController.ts]: This is a good point. I was thinking about this wrong. I was trying to use Mongo for persisted errors because I don't think we are persisting our logging to a file or anything at the moment are we?

File: src/StorageWriter/ClaimController.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (kennylavender) [reply src/StorageWriter/ClaimController.ts]: Your comment below makes a good point though, this should be a logger error that gets persisted, not added to the database.

File: src/StorageWriter/DatabaseMongo.ts (hunk 16 above)
Comment (lautarodragan) [reply src/StorageWriter/DatabaseMongo.ts]: Yep, one DAO per collection sounds best, but at the level of complexity we're at it doesn't really make any difference. 

Either way I'm fine, but I do feel strongly about not calling them "Database" or "Mongo" or "DatabaseMongo". I'd like to continue using DAO, mostly because it is (or used to be) the standard, but def open to suggestions.

File: src/StorageWriter/ClaimController.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (lautarodragan) [reply src/StorageWriter/ClaimController.ts]: Log persistence is not a responsibility of the application emitting the logs, but it WILL be done.

Applications should just write to stdout. When running in a Docker container, Docker will grab all the stdout and store it in a file for you. Another application can grab those logs and send them somewhere else.

Logz.io will handle everything for us for now, including persistence. Elastic Stack in the future, maybe.

So yeah, if that was the main reason to store them in the DB, then there's no need.

File: src/StorageWriter/ClaimController.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (lautarodragan) [reply src/StorageWriter/ClaimController.ts]: Awesome! Responded about this on the other thread.

> I was thinking it would be better to track each failure because it might fail for different reasons and it would be good to see each attempt for debugging.

If you feel there could be actual errors happening here, then yes, let's keep the log at error level and adjust if we see too many false negatives.

**Comment by kennylavender** [comment]
Todo
- [ ] Rename and Split Database into DOA's one per collection
- [ ] Log Errors instead of storing in DB.
- [ ] use custom errors

File: src/StorageWriter/ClaimController.ts [hunk 19]
```diff
@@ -1,51 +1,102 @@
...
+      return {
+        ...data,
+        ipfsFileHash,
+      }
+    } catch (error) {
+      await this.handleStoreClaimError(error, claim)
+    }
+  }
+
+  private readonly handleAddIPFSHashToClaimError = async (error: Error, claim: Claim, hash: string) => {
```
Comment (wzalazar) [inline src/StorageWriter/ClaimController.ts]: it's necessary to add `handle` word?

File: src/StorageWriter/Database.test.ts [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+  const { assert } = should('return the correct value')
+  {
+    const response: FindAndModifyWriteOpResultObject = {}
+    assert({
+      given: 'a response that does not contain a value',
+      actual: getClaimFromFindAndUpdateResponse(response),
+      expected: undefined
+    })
+  }
+  {
```
Comment (wzalazar) [inline src/StorageWriter/Database.test.ts]: Can you add an empty line before?

**Comment by kennylavender** [comment]
Blocked by https://github.com/poetapp/tslint-rules/pull/129

**Comment by kennylavender** [comment]
@lautarodragan @wzalazar ready for another round of reviews.

I also moved the configuration interfaces for the classes to the class file as discussed. It helped me manage the conflicts with master a bit easier. I can extract this change though if you would like.

File: src/StorageWriter/ClaimController.ts [hunk 21]
```diff
@@ -1,54 +1,83 @@
...
+  }
+
+  public readonly create = async (claim: Claim): Promise<void> => {
     const logger = this.logger.child({ method: 'create' })
 
-    logger.trace({ claim }, 'Storing Claim')
+    logger.trace({ claim }, 'Adding Claim')
 
-    const ipfsFileHash = await this.ipfs.addText(JSON.stringify(claim))
+    await this.daoClaims.addClaim(claim)
```
Comment (wzalazar) [inline src/StorageWriter/ClaimController.ts]: would be useful to add a `logger.trace` when finish to add the claim?

File: src/StorageWriter/DAOClaims.ts [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+  public readonly addClaim = async (claim: Claim) => {
+    await this.collection.insertOne({ claim, storageAttempts: 0, ipfsFileHash: null })
+  }
+
+  public readonly addClaimHash = async (claimId: string, ipfsFileHash: string) => {
+    await this.collection.updateOne({ 'claim.id': claimId }, { $set: { ipfsFileHash } })
+  }
+
+  private readonly findClaimToStore = () =>
+    this.collection.findOneAndUpdate(
```
Comment (wzalazar) [inline src/StorageWriter/DAOClaims.ts]: why you omit async/await in this case?

File: src/StorageWriter/StorageWriter.ts [hunk 23]
```diff
@@ -42,21 +56,31 @@ export class StorageWriter {
...
+    await this.service.start()
+
+    this.daoClaims = this.container.get('DAOClaims')
+    await this.daoClaims.start()
 
     this.logger.info('StorageWriter Started')
   }
 
   async stop() {
+    this.logger.info('Stopping StorageWriter Service')
```
Comment (wzalazar) [inline src/StorageWriter/StorageWriter.ts]: what do you say if we add an empty line per action?

File: src/StorageWriter/DAOClaims.ts (hunk 22 above)
Comment (kennylavender) [reply src/StorageWriter/DAOClaims.ts]: typescript infers the type automatically when a promise is returned. The other functions are not returning a value ( on purpose ) so they require async/await.

File: src/StorageWriter/StorageWriter.ts (hunk 23 above)
Comment (kennylavender) [reply src/StorageWriter/StorageWriter.ts]: Sounds good, adjusted, I think I understood correctly.

File: src/StorageWriter/ClaimController.ts (hunk 21 above)
Comment (kennylavender) [reply src/StorageWriter/ClaimController.ts]: added

File: src/StorageWriter/StorageWriter.ts [hunk 24]
```diff
@@ -42,21 +56,33 @@ export class StorageWriter {
...
+    await this.service.stop()
+
     this.logger.info('Stopping StorageWriter...')
     await this.router.stop()
-    this.logger.info('Stopping StorageWriter Database...')
+
+    this.logger.info('Stopping StorageWriter DAOClaims...')
```
Comment (lautarodragan) [inline src/StorageWriter/StorageWriter.ts]: Why the change from `Stopping StorageWriter Database` to `Stopping StorageWriter DAOClaims`?

Comment (kennylavender) [reply src/StorageWriter/StorageWriter.ts]: Fixed, thx. I don't remember, either find/replace or I was thinking I was making mongo connection in the DAOClaims.
